lib/psych/visitors/yaml_tree.rb

```diff
@@ -310,19 +310,20 @@ def visit_String o
           style = Nodes::Scalar::LITERAL
           plain = false
           quote = false
-        elsif o =~ /\n/
-          style = Nodes::Scalar::LITERAL
         elsif o == '<<'
           style = Nodes::Scalar::SINGLE_QUOTED
           tag   = 'tag:yaml.org,2002:str'
           plain = false
           quote = false
+        elsif o =~ /\n[^\Z]/  # match \n except blank line at the end of string
+          style = Nodes::Scalar::LITERAL
+        elsif o.length > line_width
```

Comment:
I’ve replaced `Float::INFINITY` with 0 and comparison for zero, now both cases are significantly faster than your lambda. :smirk_cat:

```ruby
require 'benchmark/ips'

str = 'Lorem ipsum dolor'
@options_false = {}
@options_true  = {line_width: 80}

def line_width_false
  @line_width_false ||= (@options_false[:line_width] || Float::INFINITY)
end

def line_width_false_zero
  @line_width_false_zero ||= (@options_false[:line_width] || 0)
end

def line_width_true
  @line_width_true ||= (@options_true[:line_width] || Float::INFINITY)
end

def line_width_true_zero
  @line_width_true_zero ||= (@options_true[:line_width] || 0)
end

wider_than_line_false = lambda { |_| false }
wider_than_line_true  = lambda { |o|  str.length > line_width_true }
wider_than_line_true_zero  = lambda { |o|  line_width_true_zero && str.length > line_width_true_zero }

Benchmark.ips do |x|
  x.report('variable:false') do
    if str.length > line_width_false
      a = 1
    end
  end
  
  x.report('variable:false-zero') do
    if line_width_false_zero != 0 && str.length > line_width_false_zero
      a = 1
    end
  end

  x.report('lambda:false') do
    if wider_than_line_false[str]
      a = 1
    end
  end


  x.report('variable:true') do
    if str.length > line_width_true
      a = 1
    end
  end
  
  x.report('variable:true-zero') do
    if line_width_false_zero != 0 && str.length > line_width_true_zero
      a = 1
    end
  end

  x.report('lambda:true') do
    if wider_than_line_true[str]
      a = 1
    end
  end

  x.report('lambda:true-zero') do
    if wider_than_line_true_zero[str]
      a = 1
    end
  end
end
```

```sh
Calculating -------------------------------------
      variable:false    71.546k i/100ms
 variable:false-zero    76.958k i/100ms
        lambda:false    78.478k i/100ms
       variable:true    78.431k i/100ms
  variable:true-zero    77.238k i/100ms
         lambda:true    66.782k i/100ms
    lambda:true-zero    62.260k i/100ms
-------------------------------------------------
      variable:false      3.981M (± 4.7%) i/s -     19.890M
 variable:false-zero      4.627M (± 4.8%) i/s -     23.087M
        lambda:false      3.972M (± 5.7%) i/s -     19.776M
       variable:true      4.520M (± 7.3%) i/s -     22.431M
  variable:true-zero      4.654M (± 4.6%) i/s -     23.249M
         lambda:true      2.839M (± 4.5%) i/s -     14.158M
    lambda:true-zero      2.112M (± 4.4%) i/s -     10.584M
```